applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp

```diff
@@ -953,6 +953,69 @@ void TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::AssignRightHand
     }
 }
 
+template <int TDim, int TNumNodes>
+BoundedVector<double, TNumNodes + 1> TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::AssembleDensityDerivativeAndShapeFunctions(
+    const double densityDerivativeWRTVelocitySquared,
+    const double densityDerivativeWRTUpwindVelocitySquared,
+    const array_1d<double, TDim>& velocity,
+    const array_1d<double, TDim>& upwindVelocity,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    const GeometryType& r_geom = this->GetGeometry();
+    const GeometryType& r_upwind_geom = pGetUpwindElement()->GetGeometry();
+
+    const array_1d<size_t, TNumNodes> upwind_node_key = GetAssemblyKey(r_geom, r_upwind_geom);
+
+    ElementalData<TNumNodes, TDim> currentElementdata;
+    ElementalData<TNumNodes, TDim> upwindElementdata;
+
+    GeometryUtils::CalculateGeometryData(r_geom, currentElementdata.DN_DX, currentElementdata.N, currentElementdata.vol);
+    GeometryUtils::CalculateGeometryData(r_upwind_geom, upwindElementdata.DN_DX, upwindElementdata.N, upwindElementdata.vol);
+
+    const BoundedVector<double, TNumNodes> current_DNV = prod(currentElementdata.DN_DX, velocity);
+    const BoundedVector<double, TNumNodes> upwind_DNV = prod(upwindElementdata.DN_DX, upwindVelocity);
+
+    BoundedVector<double, TNumNodes + 1> assembly_DNV;
+
+    for (int i = 0; i < TNumNodes; i++)
+    {
+        assembly_DNV[i] += current_DNV[i];
+        assembly_DNV[upwind_node_key[i]] += upwind_DNV[i];
+    }
+
+    return assembly_DNV;
+
+}
+
+template <int TDim, int TNumNodes>
+array_1d<size_t, TNumNodes> TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::GetAssemblyKey(
+    const GeometryType& rGeom,
+    const GeometryType& rUpwindGeom)
+{
+    std::vector<size_t> current_element_ids;
+    current_element_ids.resize(TNumNodes, false);
+    array_1d<size_t, TNumNodes> upwind_node_key;
+
+    for (unsigned int i = 0; i < TNumNodes; i++)
+    {
+        current_element_ids[i] = rGeom[i].GetDof(VELOCITY_POTENTIAL).EquationId();
+    }
+
+    for (int i = 0; i < TNumNodes; i++) { 
+        auto current_id = std::find(current_element_ids.begin(), current_element_ids.end(),
+                rUpwindGeom[i].GetDof(VELOCITY_POTENTIAL).EquationId());
```

Comment:
I think we still have the same issue, I meant something like this:

```suggestion
    std::vector<size_t> upwind_element_ids;
    pGetUpwindElement()->EquationIdVector(upwind_element_ids, CurrentProcessInfo);
    
    for (int i = 0; i < TNumNodes; i++) { 
        auto current_id = std::find(current_element_ids.begin(), current_element_ids.end(),
                upwind_element_ids[i]);
```

Does it make sense?

Reply:
The issue I had with calling the EquationIdVector function was that it requires a non-const process info variable

Reply:
I think it is still ok, I would pay de price of having a non const process info